informer: fix testcase races with hybrid overlay event handlers

dcbw opened the pull request:
There was no way to wait until the hybrid overlay informer event handlers
had stopped processing events. Narrow the race window by ensuring that
the informers themselves are done processing events before continuing
to the next test. Otherwise we hit testcase races where an informer for
a previous testcase hasn't finished yet and is accessing data that was
modified by a subsequent testcase's BeforeEach().

See https://github.com/ovn-org/ovn-kubernetes/pull/1512#issuecomment-664414658

@dave-tucker @aojea

dave-tucker commented on go-controller/hybrid-overlay/cmd/hybrid-overlay-node/hybrid-overlay-node.go:
```diff
@@ -103,12 +104,7 @@ func runHybridOverlay(ctx *cli.Context) error {
-			klog.Error(err)
-		}
-	}()
+	n.Run(stopChan, &sync.WaitGroup{})
```
Moving this outside of a goroutine stops cancellation via signal from working properly

dave-tucker commented on go-controller/hybrid-overlay/cmd/hybrid-overlay-node/hybrid-overlay-node.go:
```diff
@@ -103,12 +104,7 @@ func runHybridOverlay(ctx *cli.Context) error {
+	n.Run(stopChan, &sync.WaitGroup{})
 
 	// run until cancelled
 	<-ctx.Context.Done()
```
The `WaitGroup` shouldn't be anonymous, we should define it `wg := &sync.WaitGroup{}` and wait on it after the signal handling logic. `wg.Wait()`

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/master.go:
```diff
@@ -130,31 +131,32 @@ func NewMaster(kube kube.Interface,
 	klog.Info("Starting Hybrid Overlay Master Controller")
 
-	klog.Info("Starting workers")
+	done.Add(3)
```
nit. I would `done.Add(1)` before each goroutine instead because if someone copies/pastes the goroutine code to add a new informer, it's likely that they'll miss incrementing this here. Otherwise it probably warrants a comment.

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/master.go:
```diff
@@ -130,31 +131,32 @@ func NewMaster(kube kube.Interface,
 
 // Run starts the controller
-func (m *MasterController) Run(stopCh <-chan struct{}) {
+func (m *MasterController) Run(stopCh <-chan struct{}, done *sync.WaitGroup) {
```
nit. I would call this `wg` instead of `done`

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/master.go:
```diff
@@ -130,31 +131,32 @@ func NewMaster(kube kube.Interface,
 		if err != nil {
 			klog.Error(err)
 		}
+		done.Done()
```
nit. I would `defer wg.Done()` at the start of this goroutine.

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/master_test.go:
```diff
@@ -70,6 +71,7 @@ var _ = Describe("Hybrid SDN Master Operations", func() {
 	var (
 		app      *cli.App
 		stopChan chan struct{}
+		done     *sync.WaitGroup
```
again, `wg` would be a better name

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/node.go:
```diff
@@ -129,31 +130,35 @@ func NewNode(
 	klog.Info("Starting Hybrid Overlay Node Controller")
 
 	klog.Info("Starting workers")
+	done.Add(2)
```
Per earlier comment, I'd do this before each goroutine.

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/node.go:
```diff
@@ -129,31 +130,35 @@ func NewNode(
 
 // Run starts the controller
-func (n *Node) Run(stopCh <-chan struct{}) error {
+func (n *Node) Run(stopCh <-chan struct{}, done *sync.WaitGroup) {
```
`wg`

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/node.go:
```diff
@@ -129,31 +130,35 @@ func NewNode(
 	}()
 
-	go n.controller.RunFlowSync(stopCh)
+	done.Add(1)
```
Either leave as is, or fold this in to the one above

dcbw replied:
I'd like to leave as-is since it's not an event handler.

dave-tucker commented on go-controller/pkg/informer/informer.go:
```diff
@@ -124,7 +125,7 @@ func (e *eventHandler) Synced() bool {
 // It will block until stopCh is closed, at which point it will shutdown
 // the workqueue and wait for workers to finish processing their current work items.
-func (e *eventHandler) Run(threadiness int, stopCh <-chan struct{}) error {
+func (e *eventHandler) Run(threadiness int, stopCh <-chan struct{}, done *sync.WaitGroup) error {
```
`wg`

dave-tucker commented on go-controller/pkg/informer/informer.go:
```diff
@@ -20,7 +21,7 @@ import (
 // Deletes use the provided delete function
 type EventHandler interface {
-	Run(threadiness int, stopChan <-chan struct{}) error
+	Run(threadiness int, stopChan <-chan struct{}, done *sync.WaitGroup) error
```
`wg`

dave-tucker commented on go-controller/pkg/informer/informer_test.go:
```diff
@@ -91,7 +92,7 @@ var _ = Describe("Informer Event Handler Tests", func() {
 
 		f.Start(stopChan)
-		go e.Run(1, stopChan)
+		go e.Run(1, stopChan, &sync.WaitGroup{})
```
You should define `wg` here and ensure we `wg.Wait()` in the test.

dcbw replied:
Done, and added BeforeEach/AfterEach to handle setup/cancellation of the channel and waitgorup

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/master.go:
```diff
@@ -130,31 +131,32 @@ func NewMaster(kube kube.Interface,
 		}
+		done.Done()
 	}()
-	<-stopCh
```
Why are we not waiting on the stop channel?

dave-tucker commented on go-controller/hybrid-overlay/pkg/controller/node.go:
```diff
@@ -129,31 +130,35 @@ func NewNode(
 
 	klog.Info("Started workers")
 	n.ready = true
-	<-stopCh
```
Again, why are we not waiting on the stop channel?

dave-tucker commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -85,6 +86,8 @@ type Controller struct {
 	joinSubnetAllocator     *subnetallocator.SubnetAllocator
 	nodeLocalNatIPAllocator *ipallocator.Range
 
+	hoMaster *hocontroller.MasterController
```
Why does this need to be a member?

dcbw replied:
@dave-tucker because I move it's Run() from StartClusterMaster() to the Controller's Run() function, so we need to keep that pointer across function calls.

dave-tucker commented:
@dcbw going to need a bit more context here. Some of the changes you've made require more explanation than what's in the commit message. Specifically why the contract of `Run` has changed blocking until the channel is closed to now being non-blocking.

dcbw commented:
@dave-tucker core problem is, even when we close the stop channel, how do we know that everything that uses the stop channel as a synchronization point has finished? The stop channel is just a signal to say "stop!" but it's not an indication of completion, even if it's waited upon.

I changed the contract of Run() in a couple places to remove redundant waiting; we don't need both done.Wait() and <-stopChan. But I realize there is a consistency issue there, so I'm open to making the various Run()s block on <-stopChan while also doing the done.Wait() in the testcases.

dcbw commented:
@dave-tucker how about now? I reverted to having all Run()-type functions that previously blocked still block, but also block waiting on an internal waitgroup.

The issue I had with all the Run()-type functions that don't return until they are done is that IMO they shouldn't block on a waitgroup that has Add()s from outside the Run()-type function's own callchain. So if the thing that creates the top-level waitgroup is going to do the blocking itself, then why have sub-functions (like Run() etc) block internally, redundantly. But all that leads to a ton of:

```
wg.Add(1)
go func() {
    defer wg.Done()
    do.Something()
}
```

and multiple waitgroups. But at least all the waitgroups are scoped to their own function and I guess that's more understandable.

Anyway, addressed all your comments.

aojea commented:
> The issue I had with all the Run()-type functions that don't return until they are done is that IMO they shouldn't block on a waitgroup that has Add()s from outside the Run()-type function's own callchain

why don't make Run() return a channel instead of passing the syncGroup?

func Run(stopCh <-chan struct{}) ( <-chan error)

dave-tucker commented:
Re: @aojea's suggestion.

```
wg := &sync.WaitGroup{}
wg.Add(1)
go func() {
		defer wg.Done()
		n.Run(stopChan)
}()

<-ctx.Context.Done()
wg.Wait()
```

Would become:

```
errCh = n.Run(stopChan)

<-ctx.Context.Done()

err := <-errCh
if err != nil {
    klog.Error(err)
}
```

I think I like that API better, and it saves a goroutine.

dcbw commented:
RE returning channels instead of using goroutines, I'd prefer to keep it as-is for now as using channels is a pretty big change. But worthwhile to look at.

aojea commented:
/lgtm

danwinship commented on go-controller/hybrid-overlay/pkg/controller/node.go:
```diff
@@ -128,32 +129,44 @@ func NewNode(
-// Run starts the controller
-func (n *Node) Run(stopCh <-chan struct{}) error {
+// Run starts the controller and does not return until all operations have
+// terminated after closing the stop channel
```
"after the stop channel is closed"
(ie, it does not close the stop channel, it waits for someone else to do it)

dcbw replied:
fixed

danwinship commented on go-controller/pkg/informer/informer_test.go:
```diff
@@ -87,11 +102,12 @@ var _ = Describe("Informer Event Handler Tests", func() {
+		go func() {
+			defer wg.Done()
+			e.Run(1, stopChan)
+		}()
```
you really ought to be able to say `wg.Go(func() { e.Run(1, stopChan) })`...
(ah, but it's already been proposed and rejected)

dcbw replied:
Yeah :(

danwinship commented on go-controller/pkg/node/node.go:
```diff
@@ -244,11 +245,10 @@ func (n *OvnNode) Start() error {
-			err := nodeController.Run(n.stopChan)
-			if err != nil {
-				klog.Error(err)
-			}
```
any reason you dropped the error logging?

dcbw replied:
@danwinship yeah, after Dave Tucker's workqueue changes that function always returns `nil` so the return value was pointelss.

danwinship replied:
oh, maybe get rid of the return value too then

dcbw replied:
Yeah, I did.

dcbw commented:
@danwinship @dave-tucker updated/rebased. What did you mean about the commit summary Dan?

danwinship commented:
The commit message makes it sound like this is a small change that only involves the hybrid overlay informers... I hadn't looked at the original commit but I assumed from the conversation with Dave that this was because the original commit was much smaller and it had just grown to the point where the original message was no longer accurate. But I guess actually even the original commit was pretty big.

Anyway, the commit still only talks about hybrid overlay, but I assume it at least theoretically fixes race conditions in the master tests too?

dcbw commented:
> Anyway, the commit still only talks about hybrid overlay, but I assume it at least theoretically fixes race conditions in the master tests too?

@danwinship mostly because some of them start the hybrid overlay master code. It's still all about hybrid overlay, because that's the only code that currently uses informers.

dcbw commented:
@danwinship @trozet @dave-tucker @squeed PTAL thanks

squeed commented:
seems reasonable. Ultimately, this is really just a testing question, since we shouldn't be shutting down informers in production.
